buffer: coerce offset using Math.trunc()

cjihrig opened the pull request:
##### Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] `make -j8 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows commit guidelines
##### Affected core subsystem(s)

buffer
##### Description of change

This is a partial revert of 14d1a8a631a58408f1e85365a5f2aaa047cc0a5f, which coerced the `offset` of `Buffer#slice()` using the `|` operator. This causes some edge cases to be handled incorrectly. This commit restores the old behavior, but converts offsets to integers using `Math.trunc()`. This commit does not revert any tests, and adds an additional
regression test.

Refs: https://github.com/nodejs/node/pull/9101
Refs: https://github.com/nodejs/node/issues/9096

cc: @nodejs/buffer

Fishrock123 commented:
Does `offset >> 0` fix these bugs rather than `Math.floor()`?

cjihrig commented:
Did you mean `offset >>> 0`? If so, that would create an uint32, and `slice()` supports negative numbers.

Fishrock123 commented:
@cjihrig no, I mean a signed shift. It does similar to what `| 0` does but perhaps with less caveats.

thefourtheye commented on lib/buffer.js:
```diff
@@ -807,8 +807,8 @@ Buffer.prototype.toJSON = function() {
-  offset |= 0;
-  if (offset === 0) {
+  offset = Math.floor(+offset);
+  if (offset === 0 || Number.isNaN(offset)) {
```
Wouldn't it be better if we had `Number.isInteger` here and drop the `Math.floor` in the previous line?

cjihrig replied:
So, for example, an offset of 3.3 would become 0? That seems like it would be a surprising breaking change.

trevnorris replied:
Since `slice()` supports negative numbers `Math.round()` must be used. A simple test for that would be
```js
const buf = Buffer.from('abc');
assert.strictEqual(buf.slice(-0.5).toString(), buf.toString();
```

cjihrig commented:
> no, I mean a signed shift.

It looks like it has the same problems. Using the example from #9101:

```
> (-(-1 >>> 0) + 1) | 0
2
> (-(-1 >>> 0) + 1) >> 0
2
```

mscdex commented:
I just noticed that this PR actually fixes [another bug](https://github.com/nodejs/node/issues/9342) not covered by the regression test in this PR. Here is a test case:

```js
const buf = Buffer.from([
  1, 29,  0,   0, 1, 143, 216, 162, 92, 254, 248, 63, 0,
  0,  0, 18, 184, 6,   0, 175,  29,  0,   8,  11,  1, 0, 0
]);
const chunk1 = Buffer.from([
  1, 29, 0, 0, 1, 143, 216, 162, 92, 254, 248, 63, 0
]);
const chunk2 = Buffer.from([
  0, 0, 18, 184, 6, 0, 175, 29, 0, 8, 11, 1, 0, 0
]);
const middle = buf.length / 2;
assert.deepStrictEqual(buf.slice(0, middle), chunk1);
assert.deepStrictEqual(buf.slice(middle), chunk2);
```

cjihrig commented:
@mscdex that code is passing for me on master already, probably because of 14d1a8a631a58408f1e85365a5f2aaa047cc0a5f.

mscdex commented:
@cjihrig It doesn't pass on v6.x though.

cjihrig commented:
I think it would be simplest to backport both together.

rvagg commented:
Might be good to add that new test case to this PR even if it passes on master. Moar regression tests!

This will be good to expedite into the next v6 patch release but it'd be preferable to get it into a v7 asap.

trevnorris commented:
@cjihrig My bad. I meant to say to use `round()` in my PR review.

cjihrig commented:
Using `round()` seems like it could lead to some confusion. What about `Math.trunc()` to just drop everything after the decimal point?

trevnorris commented:
@cjihrig thanks for the correction. `Math.trunc()` is the correct solution.

cjihrig commented:
OK, moved from `Math.floor()` to `Math.trunc()`, and added the code from https://github.com/nodejs/node/pull/9341#discussion_r85653028 and https://github.com/nodejs/node/pull/9341#issuecomment-256998677 as regression tests.

mscdex commented on lib/buffer.js:
```diff
@@ -807,8 +807,8 @@ Buffer.prototype.toJSON = function() {
 function adjustOffset(offset, length) {
-  offset |= 0;
-  if (offset === 0) {
+  offset = Math.trunc(+offset);
```
Is the `+` here still necessary? It seems `Math.trunc()` handles conversions already?

cjihrig replied:
I think you're right. Removed the `+`

mscdex commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/4763/

cjihrig commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/4764/

jasnell commented:
given that use of `|` is pretty common already, would you mind adding a code comment indicating why `Math.trunc()` is used so that someone later on doesn't come along and change it? Otherwise LGTM

cjihrig commented:
@jasnell I added a comment. Does it work for you?

cjihrig commented:
Looks like there were two simultaneous CI runs today. One was green. The other had a single Jenkins issue.

jasnell commented:
Yep! Thank you!

cjihrig commented:
@trevnorris @mscdex @rvagg are any of you interested in signing off on this?

mscdex commented:
LGTM

trevnorris commented:
LGTM 

@jasnell we use binary conversions in several places we shouldn't. Should probably be fixed at some point.

jasnell commented:
Agreed

On Friday, November 4, 2016, Trevor Norris <notifications@github.com> wrote:

> LGTM
>
> @jasnell <https://github.com/jasnell> we use binary conversions in
> several places we shouldn't. Should probably be fixed at some point.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/9341#issuecomment-258502471>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAa2eazC9UGhivwqia5zjnVbztBtippKks5q63DugaJpZM4Kjgw5>
> .
>
